### PR TITLE
Issue #43: add a column for the lastest activity in a PR

### PR DESCRIPTION
**Pull request by calvarez-geny**
This PR adds a new column "Latest Activity" to the web page.

The value will be of the format "user (activity)", where activity is a short (one-word?) description of the activity.

Examples:
```calvarez-geny (committed)```
```CedricCabessa (commented)```

The types of events come from three requests ( :disappointed: ):
* the request on all the comments, which we were doing already
* a request on issue events
* a request on pr commits

The latter two are new requests.  To minimize the impact on performance, I make sure to only ask for the last single result, by adding ```per_page=1``` to the request, and retrieving the ```last``` url only.

### CODE REVIEW

File: gm_pr/practivity.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    last_commit_json = paginablejson.PaginableJson(commit_url)
+    if len(last_commit_json) == 0:
+        return None
+    last_commit_json = last_commit_json.get_last()
+    return PrActivity(dateparse.parse_datetime(last_commit_json['commit']['committer']['date']),
+                             last_commit_json['committer']['login'],
+                             "committed")
+
+# Return the PrActivity which is the most recent of the two given activities.
+def get_latest_activity(activity1, activity2):
```
Comment (CedricCabessa) [inline gm_pr/practivity.py]: may be we can use https://docs.python.org/3.5/reference/datamodel.html#object.__lt__

**Comment by CedricCabessa** [comment]
Do you know what is the perf impact? Can you measure it before / after on a typical project?

**Comment by calvarez-geny** [comment]
Screenshot.  It's the second column.

gm-pr-latest-activity

File: gm_pr/practivity.py (hunk 1 above)
Comment (calvarez-geny) [inline gm_pr/practivity.py]: looks like a good idea.  I'll give it a shot :)

File: gm_pr/practivity.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+* a commit
+* an event (examples include assigning, locking, labeling, milestoning):
+  https://developer.github.com/v3/issues/events/
+"""
+import functools
+from gm_pr import models, paginablejson, settings
+from django.utils import dateparse
+
+
+@functools.total_ordering
```
Comment (calvarez-geny) [inline gm_pr/practivity.py]: With ```functools.total_ordering```, I only need to implement ```__eq__``` and one other comparison operator: https://docs.python.org/2/library/functools.html#functools.total_ordering

File: gm_pr/practivity.py (hunk 1 above)
Comment (calvarez-geny) [inline gm_pr/practivity.py]: Implemented ```__eq__``` and ```__lt__```, but I still had to keep this method, as far as I could see:
``` 
# Return the PrActivity which is the most recent of the two given activities.
def get_latest_activity(activity1, activity2):
```
to manage the "None" case.  Let me know if I've missed something.

**Comment by calvarez-geny** [comment]
As for perf, I'll take a look.
Worst case, if these additional requests for events and commits are adding a perf hit, we can still keep the column and just put the latest commenter there (since we're already doing the comment request).

**Comment by calvarez-geny** [comment]
Perf: I added logs at the beginning and end of ```get_latest_event```and ```get_latest_commit```.  The time I measured is the time to do the query and create the PrActivity object.

In this project:
* get_latest_event:
 - 17 calls
 - Average time per request: 0.76s
* get_latest_commit:
 - 18 calls
 - Average time per request: 0.53s

I guess this is too slow to use :(  

I'll do tests for the global page loading time with and without the "latest comment" part.

**Comment by calvarez-geny** [comment]
Perf: if we keep just the latest comment (and skip the latest events and latest commits), there's no perf impact.
Test:
```time wget "http://localhost:8000/?project=My%20fav%20Google%20projects"```

Results:
* Base commit:
 - 9 tests
 - 27.09 page loading time on average
* This branch, but without events and commits:
 - 9 tests
 - 27.16 page loading time on average, which is 0.3% slower

File: bot/tasks.py [hunk 3]
```diff
@@ -46,11 +46,14 @@ def slack(url, org, weburl, repos, slackurl, channel):
                     txt += " *%s* -" % (pr.milestone)
                 for label in pr.labels:
                     txt += " *%s* -" % (label['name'])
-                txt += " %s review:%d %s:%d %s:%d %s:%d\n" % \
+                txt += " %s review:%d %s:%d %s:%d" % \
```
Comment (calvarez-geny) [inline bot/tasks.py]: PR not created properly: this change should not be part of this PR.
